src/Parallel/Reduction.hpp

```diff
@@ -235,27 +237,67 @@ void ReductionData<ReductionDatum<Ts, InvokeCombines, InvokeFinals,
 }
 /// \endcond
 
+/// Can be used instead of a `Parallel::Section` when no section is desired.
+///
+/// \see Parallel::contribute_to_reduction
+struct NoSection {};
+
 /*!
  * \ingroup ParallelGroup
  * \brief Perform a reduction from the `sender_component` (typically your own
  * parallel component) to the `target_component`, performing the `Action` upon
  * receiving the reduction.
+ *
+ * This function supports section reductions (see `Parallel::Section`). Pass
+ * the `Parallel::Section` as the \p section argument, or pass
+ * `Parallel::NoSection` to perform a reduction over the entire parallel
+ * component (default). Here's an example of a section reduction:
+ *
+ * \snippet Test_SectionReductions.cpp section_reduction
  */
-template <class Action, class SenderProxy, class TargetProxy, class... Ts>
-void contribute_to_reduction(ReductionData<Ts...> reduction_data,
-                             const SenderProxy& sender_component,
-                             const TargetProxy& target_component) noexcept {
+template <class Action, class SenderProxy, class TargetProxy, class... Ts,
+          class SectionType = NoSection>
+void contribute_to_reduction(
+    ReductionData<Ts...> reduction_data, const SenderProxy& sender_component,
+    const TargetProxy& target_component,
+    const SectionType& section = NoSection{}) noexcept {
   (void)Parallel::charmxx::RegisterReducerFunction<
       &ReductionData<Ts...>::combine>::registrar;
   CkCallback callback(
       TargetProxy::index_t::template redn_wrapper_reduction_action<
           Action, std::decay_t<ReductionData<Ts...>>>(nullptr),
       target_component);
-  sender_component.ckLocal()->contribute(
-      static_cast<int>(reduction_data.size()), reduction_data.packed().get(),
+  const auto& charm_reducer_function =
       Parallel::charmxx::charm_reducer_functions.at(
           std::hash<Parallel::charmxx::ReducerFunctions>{}(
-              &ReductionData<Ts...>::combine)),
-      callback);
+              &ReductionData<Ts...>::combine));
+  if constexpr (std::is_same_v<SectionType, NoSection>) {
+    sender_component.ckLocal()->contribute(
+        static_cast<int>(reduction_data.size()), reduction_data.packed().get(),
+        charm_reducer_function, callback);
+  } else {
+    static_assert(
+        tt::is_a_v<Section, SectionType>,
+        "Either pass a 'Parallel::Section' for the 'section' argument or "
+        "'Parallel::NoSection{}'. For the latter you can just omit the "
+        "argument.");
+    using SectionProxy = typename SectionType::cproxy_section;
+    // Retrieve the section cookie that keeps track of the reduction
+    auto& section_cookie = section.cookie();
+    // Ideally we would update the section cookie here using
```

Comment:
So this does appear to work on multiple threads _and_ nodes (currently all elliptic solves run with this code), though I expect it to break when elements migrate. As described in this code comment, I have tried to send a CkMulticast message from the array element to itself to update the cookie. However, this doesn't help because apparently the message doesn't go through CkMulticast when sent from a single element. I have opened https://github.com/UIUC-PPL/charm/issues/3351 with Charm to ask for advice. When I asked last year I think the response was that better support for section reductions was in the works, though I don't know if anything has been done yet. My current best idea for a "charm-approved" fix is to broadcast a message after load-balancing to update all cookies, but I'm curious if charm folks will suggest better ideas in the issue I linked above.

Reply:
Sorry, I'm still confused -- you don't appear to need to circulate a `CkMulticast` message to set up the cookie at the start, right? The main problem is that the cookie isn't going to be updated correctly after migration.
Can you just re-create the cookie in the same way during the migration (i.e. just re-construct it when unpacking in the pup function)? I guess I'm not understanding what the `CkMulticast` message brings with it aside from the optimization of the spanning tree that the documentation talks about.

Reply:
No I can't re-create the cookie because it keeps track of an ever-incrementing "reduction number", in addition to the multicast-related info.

Reply:
Agreed, this frustration is the topic of https://github.com/UIUC-PPL/charm/issues/3351 :)

Despite all the talk of updating section cookies in the Charm++ docs I have actually not encountered any issues so far with just not updating it. I have even done a few load-balancing tests today and it still seems to work 🤷‍♂️ I added the warning nonetheless because it's not tested thoroughly yet.